File: src/lib/scoring.ts
import { Match, Prediction, PredictedWinner } from '../types';

export function getActualWinner(match: Match): PredictedWinner | null {
  if (match.homeScore === null || match.awayScore === null) return null;
  if (match.homeScore > match.awayScore) return 'home';
  if (match.awayScore > match.homeScore) return 'away';
  return 'tie';
}

export function calculatePoints(match: Match, prediction: Prediction): number {
  if (match.homeScore === null || match.awayScore === null) return 0;

  const isExactScore =
    prediction.predictedHomeScore === match.homeScore &&
    prediction.predictedAwayScore === match.awayScore;

  if (isExactScore) return 20;

  const actualWinner = getActualWinner(match);
  if (prediction.predictedWinner === actualWinner) return 5;

  return 0;
}

export function isPredictionDeadlinePassed(matchDate: Date, now: Date = new Date()): boolean {
  // Deadline: 20 minutes before kickoff
  const deadline = new Date(matchDate.getTime() - 20 * 60 * 1000);
  return now > deadline;
}

File: sync/scoring.py
"""
Point calculation logic for predictions.
"""


def get_actual_winner(home_score: int, away_score: int) -> str:
    if home_score > away_score:
        return "home"
    elif away_score > home_score:
        return "away"
    return "tie"


def calculate_points(match_data: dict, prediction_data: dict) -> int:
    home_score = match_data.get("homeScore")
    away_score = match_data.get("awayScore")
    if home_score is None or away_score is None:
        return 0
    pred_home = prediction_data.get("predictedHomeScore", -1)
    pred_away = prediction_data.get("predictedAwayScore", -1)
    if pred_home == home_score and pred_away == away_score:
        return 20
    actual_winner = get_actual_winner(home_score, away_score)
    if prediction_data.get("predictedWinner") == actual_winner:
        return 5
    return 0
